### PR TITLE
Trainable Transformer for Slippi-AI

### PR DESCRIPTION
**Pull request by wmcnicho**
Work in progress. First pass at creating a Multi-Head Attention Block to be used in the Transformer. Added a unit test to ease debugging/development.

@vladfi1

### CODE REVIEW

File: networks.py [hunk 1]
```diff
@@ -268,6 +269,39 @@ def step(self, inputs, prev_state):
...
+    for _ in range(num_heads):
+      self.W_K.append(snt.Linear(int(output_size/num_heads))) #output is d_model/num_heads
+      self.W_V.append(snt.Linear(int(output_size/num_heads))) #output is d_model/num_heads
+      self.W_Q.append(snt.Linear(int(output_size/num_heads))) #output is d_model/num_heads
+    self.W_O = snt.Linear(output_size)
+
+  def initial_state(self, batch_size):
+    raise NotImplementedError()
+  
+  def attention(self, queries, keys, values):
```
[the remark this answers is not in the document]
Comment (vladfi1) [reply networks.py:288]: Vlad's method:

keys: [B, S, D_K]
queries: [B, S, D_K]

expanded_keys = tf.expand_dims(keys, 1): [B, 1, S, D_K]
expanded_queries = tf.expand_dims(queries, 2): [B, S, 1, D_K]

# careful: here we have broadcasting
pre_compat = expanded_queries * expanded_keys: [B, S, S, D_K]
pre_compat[b, i, j, k] = queries[b, i, k] * keys[b, j, k] 
compat = tf.reduce_sum(compat, -1): [B, S, S]

Vlad's opinion: don't do this, your way is better faster (?) and uses less memory.

File: networks.py [hunk 2]
```diff
@@ -268,6 +269,39 @@ def step(self, inputs, prev_state):
...
+    compat = queries * tf.transpose(keys) # check rank
+    norm_compat = compat / (keys.shape[-1] ** .05) # check shape
+    probs = tf.nn.softmax(norm_compat)
+    att = probs * values
+    return att
+
+  def __call__(self, queries, keys, values):
```
Comment (vladfi1) [inline networks.py:295]: Call should take `inputs: [B, S, D_M]`

(assuming only 1 head)

queries = self.W_Q(inputs): [B, S, D_K]
keys = self.W_K(keys): [B, S, D_K]
values = self.W_V(values): [B, S, D_V]

may need snt.BatchApply(self.W_Q) instead of regular self.W_Q

return attention(queries, keys, values): [B, S, D_V]

File: unit_tests.py [hunk 3]
```diff
@@ -56,5 +57,14 @@ def test_flatten_and_unflatten(self):
...
     self.assertEqual(embed_game_unflat, embed_game_struct)
 
+class Test_NetworksTransformers(unittest.TestCase):
+  def test_attention(self):
+    mhab = networks.MultiHeadAttentionBlock(8, 512)
+    # Shape grabbed from breakpoint of slippi
+    test_query = tf.ones([64, 32, 866]) # This should be the current position
+    test_keys = tf.ones([64, 32, 866]) # All in sequence
+    test_values = tf.ones([64, 32, 866]) # All in sequence
+    mhab(test_query, test_keys, test_values)
```
Comment (vladfi1) [inline unit_tests.py:67]: assert shape on output

File: networks.py [hunk 4]
```diff
@@ -268,6 +269,39 @@ def step(self, inputs, prev_state):
...
+      self.W_Q.append(snt.Linear(int(output_size/num_heads))) #output is d_model/num_heads
+    self.W_O = snt.Linear(output_size)
+
+  def initial_state(self, batch_size):
+    raise NotImplementedError()
+  
+  def attention(self, queries, keys, values):
+    compat = queries * tf.transpose(keys) # check rank
+    norm_compat = compat / (keys.shape[-1] ** .05) # check shape
+    probs = tf.nn.softmax(norm_compat)
```
Comment (vladfi1) [inline networks.py:291]: TODO: add masking so past can't depend on future

File: transformers.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    """
+    Returns a tensor following the postional encoding function
+     (sinusodal from Vaswani et. all 2017).
+
+    Return shape: [batch, seq_len, d_model]
+    """
+    def encoding_angle(pos, i):
+        pos = tf.cast(pos, tf.dtypes.float32)
+        i = tf.cast(i, tf.dtypes.float32)
+        denom = tf.map_fn(fn=lambda t: tf.math.pow(10000, t/d_model), elems=i)
```
Comment (vladfi1) [inline transformers.py:15]: Replace with `denom = tf.math.pow(10000, i/d_model)`

File: transformers.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    i_tensor = tf.expand_dims(tf.range(0, d_model), 0) # [1, d_model]
+    i_tensor = tf.repeat(i_tensor, [seq_len], axis=0) # [seq_len, d_model]
+    j_tensor = tf.expand_dims(tf.range(0, seq_len), 1) # [1 , d_model]
+    j_tensor = tf.broadcast_to(j_tensor, [seq_len, d_model]) # [seq_len, d_model]
+    #pos_tensor = tf.stack(i_tensor, j_tensor)
+    angles = encoding_angle(j_tensor, i_tensor)
+
+    # Apply sin to even indices, cos to odd indices
+    # TODO there's probably a fancy way to do this in place & save lots of memory
+    evens = tf.math.sin(angles[:, 0::2])
```
Comment (vladfi1) [inline transformers.py:27]: Apply sines and cosines to same set of d_model/2 angles, as in the paper.

File: transformers.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+
+    keys: (batch, seq_len, D_k)
+    queries: (batch, seq_len, D_k)
+    values: (batch, seq_len, D_v)
+    returns: (batch, seq_len, D_v)
+    """
+    assert keys.shape == queries.shape, "keys and values must have equivalent shapes"
+    # compat [b, i, j] is the dot product of key i and query j (for batch # b)
+    compat = tf.matmul(queries, tf.transpose(keys, [0, 2, 1])) # [B, S, S]
+    mask = tf.linalg.band_part(tf.ones((compat.shape)), -1, 0)
```
Comment (vladfi1) [inline transformers.py:58]: I would do it this way:
```
i = expand_dims(tf.range(seq_len), 1)  # [seq_len, 1]
j = expand_dims(tf.range(seq_len), 0)  # [1, seq_len]
mask = i <= j  # uses implicit broadcasting
mask[i, j] == i <= j
```

File: transformers.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+    masked_compat = compat * mask
+    if not masked:
```
Comment (vladfi1) [inline transformers.py:60]: Just put all masking operations into an if block.

Comment (vladfi1) [reply transformers.py:60]: Or if not masked, set mask to all ones matrix.

File: transformers.py [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+    """
+    Returns a tensor following the postional encoding function
+     (sinusodal from Vaswani et. all 2017).
+
+    Return shape: [batch, seq_len, d_model]
+    """
+    def encoding_angle(pos, i):
+        pos = tf.cast(pos, tf.dtypes.float32)
+        i = tf.cast(i, tf.dtypes.float32)
+        denom = tf.math.pow(10000, i/d_model)
```
Comment (vladfi1) [inline transformers.py:15]: The exponent should be `2*i/d_model`

File: networks.py [hunk 10]
```diff
@@ -268,13 +269,40 @@ def step(self, inputs, prev_state):
   def unroll(self, inputs, prev_state):
     return inputs[1], ()
 
+class TransformerWrapper(Network):
+  '''
+    A Wrapper for the output of the Encoder-only transformer, implementation in transformers.py
+  '''
+  CONFIG=dict(
+    output_size=128,
+    num_blocks=6,
```
Comment (vladfi1) [inline networks.py:278]: I'd call this `num_layers`

File: transformers.py (hunk 9 above)
Comment (vladfi1) [reply transformers.py:15]: I used this:
```
        d = tf.cast(d_model, tf.dtypes.float32)
        denom = tf.math.pow(10000., 2. * i / d)
```

Comment (wmcnicho) [reply transformers.py:15]: image

I had some trouble following this, since the LHS has an angle of `2i` i follow why they use `2i` in the RHS, but in my case I'm only using `i` so shouldn't it only be `i`?

Comment (vladfi1) [reply transformers.py:15]: No, you still want `2i`.

File: transformers.py [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+    values: (batch, seq_len, D_v)
+    returns: (batch, seq_len, D_v)
+    """
+    assert keys.shape == queries.shape, "keys and values must have equivalent shapes"
+    # compat [b, i, j] is the dot product of key i and query j (for batch # b)
+    compat = tf.matmul(queries, tf.transpose(keys, [0, 2, 1])) # [B, S, S]
+    mask = tf.ones((compat.shape))
+    if masked:
+      mask = tf.linalg.band_part(tf.ones((compat.shape)), -1, 0) # [B, S, S]
+    masked_compat = compat * mask
```
Comment (vladfi1) [inline transformers.py:59]: The masking here is wrong, it needs to be _after_ the softmax, so that the actual probabilities are zero (rather than the logits).

File: transformers.py [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+    "attention" of element j in the sequence.
+
+    keys: (batch, seq_len, D_k)
+    queries: (batch, seq_len, D_k)
+    values: (batch, seq_len, D_v)
+    returns: (batch, seq_len, D_v)
+    """
+    assert keys.shape == queries.shape, "keys and values must have equivalent shapes"
+    # compat [b, i, j] is the dot product of key i and query j (for batch # b)
+    compat = tf.matmul(queries, tf.transpose(keys, [0, 2, 1])) # [B, S, S]
```
Comment (vladfi1) [inline transformers.py:55]: Instead of explicitly calling `tf.transpose`, you can set the `transpose_b` argument for `tf.matmul` which should be faster.

File: transformers.py [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+    Returns a tensor with the same shape as Values where [b, i ,j] represents the relative importance
+    "attention" of element j in the sequence.
+
+    keys: (batch, seq_len, D_k)
+    queries: (batch, seq_len, D_k)
+    values: (batch, seq_len, D_v)
+    returns: (batch, seq_len, D_v)
+    """
+    assert keys.shape == queries.shape, "keys and values must have equivalent shapes"
+    # compat [b, i, j] is the dot product of key i and query j (for batch # b)
```
Comment (vladfi1) [inline transformers.py:54]: Great comment :)

File: transformers.py [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+    """
+    def encoding_angle(pos, i):
+        pos = tf.cast(pos, tf.dtypes.float32)
+        i = tf.cast(i, tf.dtypes.float32)
+        d = tf.cast(d_model, tf.dtypes.float32)
+        denom = tf.math.pow(10000., 2. * i/d)
+        return pos / denom
+
+    # BUG: Fails to serialize these calls:
+    # "Failed to convert object of type <class 'list'> to Tensor. Contents: [None, 480]. Consider casting elements to a supported type."
```
Comment (wmcnicho) [inline transformers.py:21]: @vladfi1 wanted to circle back on this here, I think we worked around this by disabling the compile step. Can you advise on the proper fix?

**Comment by samon11** [comment]
Hi! I'd like to help get this across the finish line. I suspect this task would really benefit from a transformer's propensity to learn long range dependencies. What is the state of this branch? Or do you think it would it be better to simply start over and possibly use TensorFlow's [transformer implementation](https://www.tensorflow.org/api_docs/python/tfm/nlp/layers/TransformerDecoderBlock)?

**Comment by wmcnicho** [comment]
Hey @samon11 awesome to hear there's some interest in this. It is quite a stale branch but at the time when we were working on it was running and learning with some success. A more established transformer implementation might be have better overall results but I'm not sure how compatible it would be.

What do you think @vladfi1?

**Comment by vladfi1** [comment]
Hey guys, sorry for not replying earlier. I'm not sure if there's much to be gained from a transformer; iirc in our experiments it achieved comparable loss to the "residual" lstm baseline. If a nice implementation gets PRed I wouldn't be opposed to merging it though.
